File: frontend/de/unika/ipd/grgen/ir/expr/graph/NodesFromIndexAccessMultipleFromToExpr.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 7.1
 * Copyright (C) 2003-2025 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

package de.unika.ipd.grgen.ir.expr.graph;

import java.util.Vector;

import de.unika.ipd.grgen.ir.*;
import de.unika.ipd.grgen.ir.expr.invocation.BuiltinFunctionInvocationExpr;
import de.unika.ipd.grgen.ir.pattern.IndexAccessOrdering;
import de.unika.ipd.grgen.ir.type.Type;

public class NodesFromIndexAccessMultipleFromToExpr extends BuiltinFunctionInvocationExpr
{
	private final Vector<IndexAccessOrdering> indexAccesses;

	public NodesFromIndexAccessMultipleFromToExpr(Vector<IndexAccessOrdering> indexAccesses, Type type)
	{
		super("nodes from index access multiple expression", type);
		this.indexAccesses = indexAccesses;
	}

	public Vector<IndexAccessOrdering> getIndexAccesses()
	{
		return indexAccesses;
	}

	/** @see de.unika.ipd.grgen.ir.expr.Expression#collectNeededEntities() */
	@Override
	public void collectNeededEntities(NeededEntities needs)
	{
		needs.needsGraph();
		for(IndexAccessOrdering indexAccess : indexAccesses) {
			indexAccess.collectNeededEntities(needs);
		}
	}
}

File: frontend/de/unika/ipd/grgen/ir/expr/graph/NodesFromIndexAccessMultipleFromToExpr.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

namespace de.unika.ipd.grgen.ir.expr.graph
{

	using System.Collections.Generic;

	using de.unika.ipd.grgen.ir;
	using BuiltinFunctionInvocationExpr = de.unika.ipd.grgen.ir.expr.invocation.BuiltinFunctionInvocationExpr;
	using IndexAccessOrdering = de.unika.ipd.grgen.ir.pattern.IndexAccessOrdering;
	using Type = de.unika.ipd.grgen.ir.type.Type;

	public class NodesFromIndexAccessMultipleFromToExpr : BuiltinFunctionInvocationExpr
	{
		private readonly IList<IndexAccessOrdering> indexAccesses;

		public NodesFromIndexAccessMultipleFromToExpr(IList<IndexAccessOrdering> indexAccesses, Type type)
			: base("nodes from index access multiple expression", type)
		{
			this.indexAccesses = indexAccesses;
		}

		public virtual IList<IndexAccessOrdering> IndexAccesses
		{
			get
			{
				return indexAccesses;
			}
		}

		/// <seealso cref="de.unika.ipd.grgen.ir.expr.Expression.collectNeededEntities() "/>
		public override void CollectNeededEntities(NeededEntities needs)
		{
			needs.NeedsGraph();
			foreach(IndexAccessOrdering indexAccess in indexAccesses)
				indexAccess.CollectNeededEntities(needs);
		}
	}

}
